### PR TITLE
Ypiel/tdi 39101 tgoogle drive get cant get google form

### PR DESCRIPTION
**Pull request by ypiel-talend**
https://jira.talendforge.org/browse/TDI-39101

### CODE REVIEW

File: main/plugins/org.talend.designer.components.localprovider/components/tGoogleDriveGet/tGoogleDriveGet_java.xml [hunk 1]
```diff
@@ -96,6 +100,46 @@
...
+	            <ITEM NAME="PNG" VALUE="image/png.png" />
+	            <ITEM NAME="SVG" VALUE="image/svg+xml.svg" />
+	        </ITEMS>
+	    </PARAMETER>
+	    
+	    <PARAMETER NAME="EXPORT_PRESENTATION" FIELD="CLOSED_LIST" NUM_ROW="130">
+	        <ITEMS DEFAULT="POWERPOINT">
+	            <ITEM NAME="POWERPOINT" VALUE="application/vnd.openxmlformats-officedocument.presentationml.presentation.pptx" />
+	            <ITEM NAME="PDF" VALUE="application/pdf.pdf"/>
+	            <ITEM NAME="OO_PRESENSTAtION" VALUE="application/vnd.oasis.opendocument.presentation.odp" />
```
Comment (gonchar-ivan) [inline main/plugins/org.talend.designer.components.localprovider/components/tGoogleDriveGet/tGoogleDriveGet_java.xml]: There is lowercase 't' in the OO_PRESENSTAtION word. Is it intentional? 
Please, fix it, if not intentional

File: main/plugins/org.talend.repository/src/main/java/org/talend/repository/model/migration/AddGoogleDocExports.java [hunk 2]
```diff
@@ -0,0 +1,124 @@
+package org.talend.repository.model.migration;
```
Comment (gonchar-ivan) [inline main/plugins/org.talend.repository/src/main/java/org/talend/repository/model/migration/AddGoogleDocExports.java]: Please, add license header.
See, https://github.com/Talend/tools/tree/master/tools-java-formatter/src/main/resources
There are files for Eclipse, which will add required templates

File: main/plugins/org.talend.repository/src/main/java/org/talend/repository/model/migration/AddGoogleDocExports.java [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+ * 
+ * @author ypiel
+ *
+ */
+public class AddGoogleDocExports extends AbstractJobMigrationTask {
+
+    /**
+     * The component to update.
+     */
+    private final static String COMPO_NAME = "tGoogleDriveGet";
```
Comment (gonchar-ivan) [inline main/plugins/org.talend.repository/src/main/java/org/talend/repository/model/migration/AddGoogleDocExports.java]: Maybe better not shorten field name. I prefer `COMPONENT_NAME`
However, it's up to you

File: main/plugins/org.talend.designer.components.localprovider/components/tGoogleDriveGet/tGoogleDriveGet_java.xml (hunk 1 above)
Comment (ypiel-talend) [inline main/plugins/org.talend.designer.components.localprovider/components/tGoogleDriveGet/tGoogleDriveGet_java.xml]: Not at all. Corrected.

File: main/plugins/org.talend.repository/src/main/java/org/talend/repository/model/migration/AddGoogleDocExports.java (hunk 2 above)
Comment (ypiel-talend) [inline main/plugins/org.talend.repository/src/main/java/org/talend/repository/model/migration/AddGoogleDocExports.java]: Thanks. It strange, my eclipse is well configured with this template in prefrences/Java/code style/code template/Comments/Files
I have already imported the xml file you have linked. When I test, the template is not included. 

... I had to tick the check box 'Automatically add comments for new methods and types', and now it's ok.

File: main/plugins/org.talend.repository/src/main/java/org/talend/repository/model/migration/AddGoogleDocExports.java (hunk 3 above)
Comment (ypiel-talend) [inline main/plugins/org.talend.repository/src/main/java/org/talend/repository/model/migration/AddGoogleDocExports.java]: I agree.

File: main/plugins/org.talend.repository/src/main/java/org/talend/repository/model/migration/AddGoogleDocExports.java (hunk 2 above)
Comment (gonchar-ivan) [inline main/plugins/org.talend.repository/src/main/java/org/talend/repository/model/migration/AddGoogleDocExports.java]: I had similar problem. In my case I set this property only for particular project. But it is possible to set this property for all projects.
- right click on project -> Properties (project specific properties)
- Window -> Preferences (common properties)

File: main/plugins/org.talend.designer.components.localprovider/components/tGoogleDriveGet/tGoogleDriveGet_begin.javajet [hunk 4]
```diff
@@ -19,6 +21,8 @@ imports="
 		String cid = node.getUniqueName();
 		NodeParamsHelper helper = new NodeParamsHelper(node);
 
+		String exportExtCacheName = "tGoogleDriveGet_exportExt";
```
Comment (gonchar-ivan) [inline main/plugins/org.talend.designer.components.localprovider/components/tGoogleDriveGet/tGoogleDriveGet_begin.javajet]: I don't see where it is used

File: main/plugins/org.talend.designer.components.localprovider/components/tGoogleDriveGet/tGoogleDriveGet_begin.javajet [hunk 5]
```diff
@@ -108,28 +127,64 @@ com.google.api.services.drive.Drive <%=cid%>_client;
...
 			com.google.api.services.drive.model.File <%=cid%>_file = <%=cid%>_client.files().get(<%=cid%>_fileId).execute();
-			com.google.api.client.http.HttpResponse resp = <%=cid%>_client.getRequestFactory().buildGetRequest(new com.google.api.client.http.GenericUrl(<%=cid%>_file.getDownloadUrl())).execute();
+			
+			String <%=cid%>_downloadUrl = <%=cid%>_file.getDownloadUrl();
+
+			// Google document can't be downloaded, download url is null, they must be exported
+			if(<%=cid%>_downloadUrl == null) {
+				java.util.Map<String, String> <%=cid%>_exportLinks = <%=cid%>_file.getExportLinks();
+				
+                if(<%=cid%>_exportLinks != null) {
```
Comment (gonchar-ivan) [inline main/plugins/org.talend.designer.components.localprovider/components/tGoogleDriveGet/tGoogleDriveGet_begin.javajet]: strange formatting here. However it is ok in some text editor.
The reason is that previous lines has `\t` indent, and these uses spaces.
Not a big problem, but better to keep indent consistent. 
(But seems, we have no standard for .javajet file. Different files have different approach. Also I don't know how format .javajet files automatically ((( )

File: main/plugins/org.talend.designer.components.localprovider/components/tGoogleDriveGet/tGoogleDriveGet_begin.javajet [hunk 6]
```diff
@@ -108,28 +127,64 @@ com.google.api.services.drive.Drive <%=cid%>_client;
 		{
 			String <%=cid%>_fileId = <%=cid%>_files.getItems().get(0).getId();
 			com.google.api.services.drive.model.File <%=cid%>_file = <%=cid%>_client.files().get(<%=cid%>_fileId).execute();
-			com.google.api.client.http.HttpResponse resp = <%=cid%>_client.getRequestFactory().buildGetRequest(new com.google.api.client.http.GenericUrl(<%=cid%>_file.getDownloadUrl())).execute();
+			
+			String <%=cid%>_downloadUrl = <%=cid%>_file.getDownloadUrl();
```
Comment (gonchar-ivan) [inline main/plugins/org.talend.designer.components.localprovider/components/tGoogleDriveGet/tGoogleDriveGet_begin.javajet]: Just a note. Seems in other components we use `<%=cid%>` as suffix more rather as prefix.
But let's keep here as it is. As I know originally GoogleDrive component was developed by another team.

File: main/plugins/org.talend.designer.components.localprovider/components/tGoogleDriveGet/tGoogleDriveGet_begin.javajet [hunk 7]
```diff
@@ -108,28 +127,64 @@ com.google.api.services.drive.Drive <%=cid%>_client;
...
+                    
+                    <%=cid%>_downloadUrl = <%=cid%>_exportLinks.get(<%=cid%>_exportTo);
+<%
+                if(addExt) {
+%>              
+                    <%=cid%>_outputFileExt = ext;
+<%
+                }
+%>
+                }
```
Comment (gonchar-ivan) [inline main/plugins/org.talend.designer.components.localprovider/components/tGoogleDriveGet/tGoogleDriveGet_begin.javajet]: What if `exportLinks` are also null? Is it possible?

File: main/plugins/org.talend.designer.components.localprovider/components/tGoogleDriveGet/tGoogleDriveGet_begin.javajet (hunk 4 above)
Comment (ypiel-talend) [inline main/plugins/org.talend.designer.components.localprovider/components/tGoogleDriveGet/tGoogleDriveGet_begin.javajet]: In my first developpment I had a map for link mime/Type with extension and I wanted to cache this map into globalMap. But, I added extension directly in mime/Type closed lists. I think it's more elegant.
Removed.

File: main/plugins/org.talend.designer.components.localprovider/components/tGoogleDriveGet/tGoogleDriveGet_begin.javajet (hunk 6 above)
Comment (ypiel-talend) [inline main/plugins/org.talend.designer.components.localprovider/components/tGoogleDriveGet/tGoogleDriveGet_begin.javajet]: Ok, I take it in account.

File: main/plugins/org.talend.designer.components.localprovider/components/tGoogleDriveGet/tGoogleDriveGet_begin.javajet (hunk 5 above)
Comment (ypiel-talend) [inline main/plugins/org.talend.designer.components.localprovider/components/tGoogleDriveGet/tGoogleDriveGet_begin.javajet]: I use the java editor of eclipse. As requested here : https://github.com/Talend/tools/tree/master/tools-java-formatter
I have imported the Talend formatter profil which use 4 spaces for tab.

it seems taht the previous developpment  has used tabs.

Should I change all tab in 4xspaces ?

Comment (gonchar-ivan) [inline main/plugins/org.talend.designer.components.localprovider/components/tGoogleDriveGet/tGoogleDriveGet_begin.javajet]: I don't know for the moment. Let's keep as it for the moment and let's raise this question on next component meeting to develop some coding conventions for .javajet

Comment (gonchar-ivan) [inline main/plugins/org.talend.designer.components.localprovider/components/tGoogleDriveGet/tGoogleDriveGet_begin.javajet]: My current strategy is to keep the same indent as which is already used in current .javajet file

File: main/plugins/org.talend.designer.components.localprovider/components/tGoogleDriveGet/tGoogleDriveGet_begin.javajet (hunk 7 above)
Comment (ypiel-talend) [inline main/plugins/org.talend.designer.components.localprovider/components/tGoogleDriveGet/tGoogleDriveGet_begin.javajet]: Yes, if `exportLinks` is null, the document can't be downloaded. It is true for google Maps/Sites/Forms...
I create a ticket to propose to add a new reject output for this component. I think I have to discuss with Pierre.
https://jira.talendforge.org/browse/TDI-39247?filter=-2

File: main/plugins/org.talend.designer.components.localprovider/components/tGoogleDriveGet/tGoogleDriveGet_begin.javajet [hunk 8]
```diff
@@ -27,9 +29,40 @@ imports="
...
 
 		boolean dieOnError = helper.getBoolParam("__DIE_ON_ERROR__");
 		
+		String exportDocumentType = helper.getStringParam("__EXPORT_DOCUMENT__");
+		String exportDrawType = helper.getStringParam("__EXPORT_DRAWING__");
+		String exportPresentationType = helper.getStringParam("__EXPORT_PRESENTATION__");
+		String exportSpreadSheetType = helper.getStringParam("__EXPORT_SPREADSHEET__");
+		
+		Map<String, String> mapExtension = new HashMap<String, String>();
+		mapExtension.put("application/epub+zip", ".zip");
```
Comment (gonchar-ivan) [inline main/plugins/org.talend.designer.components.localprovider/components/tGoogleDriveGet/tGoogleDriveGet_begin.javajet]: There is some comment about it from Pierre. We need to ask him

**Pull request merged**